### PR TITLE
Improve modem command-mode handling

### PR DESCRIPTION
**Pull request by NicknineTheEagle**
* Actually use modem registers for setting CR, LF and BS chars.
* Ignore everything in modem command mode until "AT" sequenced is encountered.

Fixes:
* The Need for Speed - sends "+++AT" to the modem at init time.
* Duke Nukem 3D - sends three ESC chars before modem init command.

### CODE REVIEW

**Comment by kcgen** [comment]
Thanks @NicknineTheEagle; these are notable games and certainly want them working.

Can you split the addition of the echo() function (and calling it) into a separate commit? 

This main reason is to make the three concepts in this PR compartmentalized easy to bisect, if needed.

If your modem regression testing is coming back positive, then will be happy to merge.

**Comment by kcgen** [comment]
(out of scope, but just mentioning it)

Regarding the thee +++'s or three escapes prior to AT:

> That's the "Escape Sequence" that switches the modem from data to command state, which usually disconnects modem.; this command is not proceeded by AT. Most modems have a "guard time" of 1 second in which no additional data should be sent to the modem before or after the escape code is sent, or it will be ignored. 

---

I'm not sure if the code current checks for the "quiet time" before and after the escape sequence is received as criteria to switch to command mode.. but that would be another welcome feature. 👍

**Comment by NicknineTheEagle** [comment]
> (out of scope, but just mentioning it)
> 
> Regarding the thee +++'s or three escapes prior to AT:
> 
> > That's the "Escape Sequence" that switches the modem from data to command state, which usually disconnects modem.; this command is not proceeded by AT. Most modems have a "guard time" of 1 second in which no additional data should be sent to the modem before or after the escape code is sent, or it will be ignored.
> 
> I'm not sure if the code current checks for the "quiet time" before and after the escape sequence is received as criteria to switch to command mode.. but that would be another welcome feature. 👍

I'm aware of what +++ is, the issue is that games often can't know for certain if modem is in command mode so they'll send +++ before firing commands. Real modem is supposed to ignore those in command mode.

**Comment by kcgen** [comment]
@NicknineTheEagle> (tested) Doom, Descent, The Need for Speed, Duke Nukem 3D. Duke 3D connects fine but crashes in-game, this issue came all the way from SVN.  This happens in UDP mode, too, which furthers my suspicions that the crash has nothing to do with modem emulation.

Thanks for elaborating on your testing @NicknineTheEagle. Your PR is working much better in 688 Attack Sub as well:

File: src/hardware/serialport/softmodem.cpp [hunk 1]
```diff
@@ -851,6 +851,15 @@ void CSerialModem::TelnetEmulation(uint8_t *data, uint32_t size)
...
 
+void CSerialModem::Echo(uint8_t ch)
+{
+	if (echo) {
+		rqueue->addb(ch);
+		// LOG_MSG("SERIAL: Port %" PRIu8 " echo back to queue: %x.",
+		//         GetPortNumber(), txval);
+	}
+}
+
```
Comment (kcgen) [inline src/hardware/serialport/softmodem.cpp:853]: :+1: - looks good - and thanks for keeping the interface "8-bit clean".

File: src/hardware/serialport/softmodem.cpp [hunk 2]
```diff
@@ -490,14 +479,16 @@ void CSerialModem::DoCommand()
...
-				        socketType ? "ENet" : "TCP");
-				// Reset port state.
-				EnterIdleState();
+				if (socketType != (SocketTypesE)requested_mode) {
+					socketType = (SocketTypesE)requested_mode;
+					// This will break when there's more than two
+					// socket types.
+					LOG_MSG("SERIAL: Port %" PRIu8 " socket type %s",
+							GetPortNumber(),
+							socketType ? "ENet" : "TCP");
```
Comment (kcgen) [inline src/hardware/serialport/softmodem.cpp:488]: For the two indented lines that are part of LOG_MSG: `GetPortNumber(),` and `socketType ...`

Can you tab them inward up to the same level as the `LOG_MSG`, and from there, space them inward until they line up with the first double-quote `"` in `"SERIAL`.

This way, muilti-line alignment for fuinction arguments will always be lined up, regardless of tab-size.

If you have clang-format, you can run `./scripts/format-commit.sh` or `git clang-format`, both of which will just apply formatting clean rules (like this) automatically to any changed lines in the current commit.  It's very handy for fixing stuff like this.  I think vscode and Visual Studio can auto-apply it too.

**Review by kcgen — CHANGES_REQUESTED** [review]
Just one minor whitespace comment; nothing else after that. 
Thank you @NicknineTheEagle !

**Review by kcgen — APPROVED** [review]
Adding to backports for 0.78.x

**Pull request merged**